src/core/coap/coap_client.cpp

```diff
@@ -308,9 +308,7 @@ Message *Client::FindRelatedRequest(const Header &aResponseHeader, const Ip6::Me
     {
         aRequestMetadata.ReadFrom(*message);
 
-        if (((aRequestMetadata.mDestinationAddress == aMessageInfo.GetPeerAddr()) ||
-             aRequestMetadata.mDestinationAddress.IsMulticast()) &&
-            (aRequestMetadata.mDestinationPort == aMessageInfo.GetPeerPort()))
+        if (aRequestMetadata.mDestinationPort == aMessageInfo.GetPeerPort())
```

Comment:
I would consider using IsAnycastRoutingLocator as an additional check for destination address instead of removing address checks.